### ISSUE TITLE
Feature-Request: PI Program Board with Red Strings

### ISSUE THREAD

**Issue by CodeFreezr** [wekan/wekan#3392]
Hello Dear Wekanist,

we use Wekan keen as a Tool for our SAFe PI Planings. We are missing a possibility to visualize dependencies at a glance.
It could be a "layer" on top, where we manual create "Red String" connections via drag and drop.

If this "drag and drop" flow goes in conrete field values would be nice, but the visual overview ist for the PI-Planing Use-Case more important.

See attached some Example-Screenshot from other tools, like 

- [piplanning.io](https://www.piplanning.io/), 
- [kendis ](https://kendis.io/scaling-agile/significance-managing-dependencies/) or an original example from 
- [scaledagileframework](https://www.scaledagileframework.com/pi-planning/).

![Dependencies-Kendis-SAFe-LeSS](https://user-images.githubusercontent.com/966869/101896971-1c49a680-3baa-11eb-8dcf-48c6a5a6b2ad.png)
![2020-12-11 12_05_36-piplanning app](https://user-images.githubusercontent.com/966869/101896985-210e5a80-3baa-11eb-8ffc-65fa70ee1b98.png)
![F3-Program-Board](https://user-images.githubusercontent.com/966869/101897001-25d30e80-3baa-11eb-87c4-8a04375c13ce.png)
![program-board](https://user-images.githubusercontent.com/966869/101897005-279cd200-3baa-11eb-88b2-a4804b1c107c.jpg)

**Comment by xet7** [comment]
@CodeFreezr 

Is any of this kind of "Red String" code Open Source?

Please email support@wekan.team with info about how much you are currently paying for these features, and what it would be worth to you to have this feature in Wekan. I'll then think do I dare to try to implement this.

**Comment by xet7** [comment]
I asked at Freenode IRC sandstorm channel, some similar to "Red String" code, for drawing on top with SVG, has been made by isd at https://github.com/zenhack/dndgrid . Client code is in `src/elm/` , and probably want to look at Lines.elm and references to StartDraw/StopDraw in Main.elm .

**Comment by CodeFreezr** [comment]
I'm not sure if parts of this "Apache v2" licensed Code could help: https://github.com/jgraph/drawio

**Comment by CodeFreezr** [comment]
This solution doesn't seem to support curved lines and seems a bit dated. Perphaps as inspiration? https://github.com/alexgheorghiu/diagramo

**Comment by CodeFreezr** [comment]
Another different solution, but with a js drawing lib inside: https://github.com/bpmn-io/bpmn-js. Very active project with an own OSS-like license.

**Comment by CodeFreezr** [comment]
If you would ask me, what is the best "Drag and Drop" UX beside Kendis and piplaning.io overall? I would say: miro app. This "Realtime- Whiteboard" Solution is really awesome, but as I have understand, the core-libs are not OSS, or? But there are a lot of (SDK) stuff around on github, perhaps some inspirations? https://github.com/miroapp

**Comment by CodeFreezr** [comment]
Ok, this above solutions do much, much more then only drawing "Red Strings". But perhaps there is some code inside we could use as a starting point?

**Comment by xet7** [comment]
Added to [Roadmap](https://boards.wekan.team/b/D2SzJKZDS4Z48yeQH/wekan-open-source-kanban-board-with-mit-license/R9eHj5MJpKTc5oWDS)

**Comment by xet7** [comment]
It seems [here are some CSS curved lines](https://stackoverflow.com/questions/27607844/drawing-curved-lines-to-connect-elements-on-web-page).

TODO: Check is something similar possible with HTML4 using NetSurf browser.

HTML:
```
<div id="upper-line"></div>

<div id="lower-line"></div>

<div id="content-1">content 1</div>

<div id="content-2">content 2</div>

<div id="content-3">content 3</div>
```
CSS:
```
#upper-line {
    border: solid 1px #000;
    width:80%;
    height: 250px;
    border-radius: 50%;
    left:55px;
    border-right: none;
    border-top: none;
    border-bottom: none;
    position: absolute;
    top: 100px;
}

#lower-line {
    border: solid 1px #000;
    width: 80%;
    height: 250px;
    border-radius: 50%;
    top: 340px;
    left: -60px;
    border-left: none;
    border-top: none;
    border-bottom: none;
    position: absolute
}

#content-1 {
    position: absolute;
    left: 180px;
    top: 75px;
    width: 100px;
    height: 100px;
    line-height: 100px;
    border-radius: 50%;
    text-align: center;
    background-color: orange;
}

#content-2 {
    position: absolute;
    left: 40px;
    top: 200px;
    width: 100px;
    height: 100px;
    line-height: 100px;
    border-radius: 50%;
    text-align: center;
    background-color: #98879A;
}

#content-3 {
    position: absolute;
    left: 400px;
    top: 400px;
    width: 100px;
    height: 100px;
    line-height: 100px;
    border-radius: 50%;
    text-align: center;
    background-color: #637DBA;
}
```

**Comment by CodeFreezr** [comment]
Here an additional Description for this Feature-Request:

Generally speaking, the requirement is this:
•	One card can have “must be started after” type dependencies to 1:n other cards.
•	These dependencies can be added and deleted in the detailed view of a card (i.e. selectable by a “search by title” functionality).
•	The dependencies are maintained and ideally synced if a card is moved or copied to different boards.
•	Whenever cards that are dependent to each other are on the same board, a visual connection is drawn between the cards
o	Ideally, the connection is always drawn from the right edge of the prerequisite card to the left edge of the dependent card to give further visual indication of which card depends on what.
o	If a card has dependencies to multiple cards on the same board, they’re all drawn.
 
Differentiation
•	Having one or multiple dependencies to cards on the same board must not limit the columns on which a card can be moved.
This is to say: If board columns signify Scrum Sprints, weeks or “phases”, it would be possible to place the prerequisite card _after_ the dependent card. We believe it would be an unnecessary complexity to enforce “good” dependency management via software and we welcome the freedom during planning to temporarily place items in any column.

**Comment by CodeFreezr** [comment]
Any news here?

**Comment by xet7** [comment]
@CodeFreezr 

Development is In Progress. When I get it working, feature will be added to Meteor WeKan or Multiverse WeKan.

Multiverse WeKan will use SQLite or some other database. It will support also HTML4 browsers like Netsurf/Lynx/Links/ELinks/w3m , Amiga webbrowsers like AWeb and iBrowse, etc with minimal HTML/CSS. Example static html page with additional background image feature is at https://wekan.team/test/

**Comment by CodeFreezr** [comment]
Thx for your Feedback. At the Moment the Meteor WeKan is ours. Did you receive our offering for a live meeting?

**Comment by CodeFreezr** [comment]
Nice to see that others see a need here too. https://github.com/wekan/wekan/issues/3860

Dominiques resends you the email. And yes, my first name is Detlef^^

**Comment by xet7** [comment]
Some progress.

I figured out how to add some svg code temporarily to board title, so it shows red line:

```
<style type="text/css">
.redlines {
  position: fixed; /* or position: absolute */
  top: 220px;
  left: 250px;
  width: 50px;
  height: 2px;
  pointer-events: none; /* so that below cards can be clicked */
}
</style>
<div class="redlines">
<svg height="210" width="500">
  <line x1="0" y1="0" x2="270" y2="150" style="stroke:rgb(255,0,0);stroke-width:2" />
  Sorry, your browser does not support inline SVG.
</svg>
</div>
```
![Redlines](https://user-images.githubusercontent.com/15545/174467670-3973019e-4505-4638-8a48-8356d9dcb96a.png)

At WeKan input fields, any Javascript is not run, it's sanitized for XSS reasons by [dompurify](https://www.npmjs.com/package/dompurify), so SVG images at WeKan board can not include Javascript inside SVG [like this SVG dragging example](https://www.petercollingridge.co.uk/tutorials/svg/interactive/dragging/). So I'll try setting positioning in WeKan code itself some way. I presume it could be similar like relative positioning when dragging cards, sorting, etc.

I found video about Red Strings from this piplanning YouTube video:
https://www.youtube.com/watch?v=RgNdEUmkQXo

That video seems to be all about realtime sync of all board members collaboration, like current Meteor WeKan.

**Comment by CodeFreezr** [comment]
Perhaps, if it is too complex to implement red strings direct into Wekan GUI, perhaps we find a way to export/sync a board, perhaps as background for e.g. into excalidraw (a better FOSS Miro-Board). Then, after finishing drawing the connection, analyse excalidraw and add "CustomField" with UID-Links and Relative Position Coords on related cards?

I do this at the moment sometimes "manual" via a sliced screenshot and than work with a "bitmap" per card on excalidraw. It's nor realy genius but it works more or lesse.

excalidraw Homepage: https://excalidraw.com/
excalidraw Upstream: https://github.com/excalidraw/excalidraw

**Comment by xet7** [comment]
@CodeFreezr 

> Thx for your Feedback. At the moment the Meteor WeKan is ours. Multiverse WeKan ist too simplified. Did you receive our offering for a live meeting?

Current Meteor WeKan GUI is too heavy. When both server and client computers have 32 GB RAM, it's too slow. 10 MB of Javascript is too much.

Only way to add even more features to WeKan is to continue developing Multiverse WeKan like UI https://wekan.team/test/ where at browserside is only HTML4/CSS, without any Javascript. That Multiverse WeKan is not simplified, it has some more features that current Meteor WeKan does not have.

> Perhaps, if it is too complex to implement red strings direct into Wekan GUI...

It is not too complex to implement red strings directly into Multiverse WeKan GUI.

**Comment by xet7** [comment]
More details about Multiverse at https://github.com/wekan/wekan/issues/4578#issuecomment-1248485306

**Comment by xet7** [comment]
1) Connected boxes examples, VML+SVG, partially works also at Internet Explorer:

- https://dmitrybaranovskiy.github.io/raphael/graffle.html
- https://dmitrybaranovskiy.github.io/raphael/
- https://github.com/DmitryBaranovskiy/raphael
- https://github.com/voidlabs/elycharts
- https://www.npmjs.com/search?q=vml
- https://en.wikipedia.org/wiki/Vector_Markup_Language

2) Raylib

- Has some bezier example https://www.raylib.com/examples.html
- https://www.raylib.com
- https://github.com/raysan5/raygui
- https://github.com/raysan5/raylib
- Native compiled example could maybe work also on RISC-V device like MangoPi, I'll try sometime

3) macOS/iOS related cross platform
- Swift related https://github.com/TokamakUI/Tokamak
- .NET Avalonia is an option which targets OSX in addition to other platforms
- https://news.ycombinator.com/item?id=33394807

4. Progressive enhancement

4a) Svelte
- https://svelte.dev
- Progressive enhancement demo here, where webpage functionality works also without Javascript https://youtu.be/3QEoJRjxnxQ?t=3107
- https://en.wikipedia.org/wiki/Progressive_enhancement

4b) wt C++ web framework, uses progressive enhancement
- has some VML too, but license GPL
- https://en.wikipedia.org/wiki/Wt_(web_toolkit)
- https://www.webtoolkit.eu/wt
- https://github.com/emweb/wt

4c) Haxe UFront
- Not currently maintained, but has some progressive enhancement features
- https://github.com/ufront/ufront

**Comment by xet7** [comment]
Moved to here from #4777

From @hw-p 

Feature Request: Whiteboard
It will be amazing to have whiteboard.

From @xet7 

Yes, when I get it working. In Progress.

**Comment by xet7** [comment]
About Apple's new app Freeform for macOS and iOS:
- If copying from Freeform directly to Inkscape, it copies as bitmap
- If exporting from Freeform to PDF, and then Importing PDF to Inksape, it is vector graphics, objects of drawing can be independently moved, resized, edited etc.

**Comment by helioguardabaxo** [comment]
Hi, @xet7 !
Is this a dependency relation between cards?
Why not create some type for a "Dependency" field to link to another card like:
- related to #
- blocks #
- is blocked by #
- fixes #
- is fixed by #

To help in user experience, would be good if I can to customize icon and color fields on minicard.
Filters by "Dependency" field would be good too.

**Comment by xet7** [comment]
@helioguardabaxo 

Kind of. Also direction of dependency relation would need to be saved, those arrows at first screenshot of first comment on this issue https://github.com/wekan/wekan/issues/3392#issue-762248381

This feature has already been paid. I will figure out how to code it.

**Comment by xet7** [comment]
Someone thinking about many lines/strings:
- https://gboisse.github.io/posts/node-graph/
- https://news.ycombinator.com/item?id=37702059

Related to dragging: 
- https://github.com/wekan/wekan/wiki/Drag-Drop#drag-code-examples

**Comment by xet7** [comment]
Related:
- https://news.ycombinator.com/item?id=38710030
- https://github.com/zsviczian/obsidian-excalidraw-plugin

**Comment by TimUnverzagt** [comment]
Hi @xet7 
Is this feature still in development?

**Comment by xet7** [comment]
@TimUnverzagt 

Yes. This feature has been paid, and development is in progress.

**Comment by xet7** [comment]
In general, every paid feature will be added to WeKan.

**Comment by xet7** [comment]
Implemented an MVP in 5bd1d0ae742e118561b0e9e3170c1e44bd579ad8 (in the next release). Card-to-card dependencies are now visualized as red, arrow-headed connection lines drawn on top of the board — a SAFe PI-planning style program board. Each card has a "Dependencies" section in its detail view to add/remove links to other cards on the same board, and a board-header toggle renders an SVG overlay that draws a red curve from each card to each of its dependencies, following the cards on scroll/resize (the overlay is non-interactive so cards stay clickable). Thanks @CodeFreezr for the request and @dbt4u for the references. Further enhancements (drag-to-connect, curved-line styling, cross-swimlane routing) can be tracked in follow-up issues.
